src/db.c

```diff
@@ -124,6 +124,11 @@ robj *lookupKey(redisDb *db, robj *key, int flags) {
         if (server.current_client && server.current_client->flags & CLIENT_NO_TOUCH &&
             server.current_client->cmd->proc != touchCommand)
             flags |= LOOKUP_NOTOUCH;
+        if (server.maxmemory != 0 && (server.maxmemory_policy & MAXMEMORY_FLAG_NO_SHARED_INTEGERS) &&
```

Comment:
it's better to abstract a function or define a macro for here and `tryObjectEncoding`.

Reply:
The problem is that `tryObjectEncoding` doesn't encode shared objects.
```
/* It's not safe to encode shared objects: shared objects can be shared
 * everywhere in the "object space" of Redis and may end in places where
 * they are not handled. We handle them only as values in the keyspace. */
 if (o->refcount > 1) return o;
```

Reply:
I mean something like this:

```c
int canUseSharedObject(void ) {
    return server.maxmemory == 0 || !(server.maxmemory_policy & MAXMEMORY_FLAG_NO_SHARED_INTEGERS);
}

robj *tryObjectEncodingEx(robj *o, int try_trim) {
...
        if (canUseSharedObject() &&
            value >= 0 &&
            value < OBJ_SHARED_INTEGERS)
        {
            decrRefCount(o);
            return shared.integers[value];
        }
}

robj *lookupKey(redisDb *db, robj *key, int flags) {
...
        if (!canUseSharedObject() && val->refcount == OBJ_SHARED_REFCOUNT) {
            val = dupStringObject(val)
        }
}

```